File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/QueryStatisticsHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.{FusedWorkerStatistics, WorkerAsyncRPCHandlerInitializer, WorkerStatistics}
import edu.uci.ics.amber.engine.architecture.worker.promisehandlers.QueryStatisticsHandler.QueryStatistics
import edu.uci.ics.amber.engine.common.ITupleSinkOperatorExecutor
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.{CommandCompleted, ControlCommand}

object QueryStatisticsHandler {
  final case class QueryStatistics() extends ControlCommand[Either[WorkerStatistics,FusedWorkerStatistics]]
}

trait QueryStatisticsHandler {
  this: WorkerAsyncRPCHandlerInitializer =>

  registerHandler { (msg: QueryStatistics, sender) =>
    val state = stateManager.getCurrentState
    val arr = dataProcessor.collectStatistics()
    if(arr.length == 1){
      val result = operator match {
        case sink: ITupleSinkOperatorExecutor =>
          Option(sink.getResultTuples())
        case _ =>
          Option.empty
      }
      Left(WorkerStatistics(state, arr.head._1, arr.head._2, result))
    }else{
      Right(FusedWorkerStatistics(state, arr))
    }
  }
}

File: core/amber/src/main/python/core/architecture/handlers/query_statistics_handler.py
from edu.uci.ics.amber.engine.architecture.worker import QueryStatistics, WorkerStatistics
from .handler_base import Handler
from ..managers.context import Context


class QueryStatisticsHandler(Handler):
    cmd = QueryStatistics

    def __call__(self, context: Context, command: QueryStatistics, *args, **kwargs):
        input_count, output_count = context.statistics_manager.get_statistics()
        state = context.state_manager.get_current_state()

        return WorkerStatistics(worker_state=state, input_row_count=input_count, output_row_count=output_count)
